### PR TITLE
[SIMD.js] Implement conversion functions of partial simd128 data types i...

### PR DESCRIPTION
**Pull request by chuan9**
...n runtime.

The APIs include:
SIMD.float32x4.fromFloat64x2
SIMD.float32x4.fromFloat64x2Bits
SIMD.float64x2.fromFloat32x4
SIMD.float64x2.fromFloat32x4Bits
SIMD.int32x4.fromFloat64x2
SIMD.int32x4.fromFloat64x2Bits
SIMD.float64x2.fromInt32x4
SIMD.float64x2.fromInt32x4Bits

### CODE REVIEW

**Comment by chuan9** [comment]
@rakuco Can you tell me which kinds of test does the bot run? debug or release or both of them? I run the make arm.release.check and the test is ok, can you give me the failure report? thanks!

**Comment by rakuco** [comment]
> @rakuco Can you tell me which kinds of test does the bot run? debug or release or both of them?

This information is available in the links with the statuses. For example, the blue text at the top of [this page](https://build.crosswalk-project.org/try/builders/V8%20Linux%20ARM/builds/98/steps/compile/logs/stdio) shows that the bot builds in Release mode. The tests are all listed [here](https://build.crosswalk-project.org/try/builders/V8%20Linux%20ARM/builds/98) along with the commands run in each of them.

> I run the make arm.release.check and the test is ok, can you give me the failure report? thanks!

It's all in the links above. In this case though, I believe it failed because GitHub was suffering a days-long DDoS attack (test262 failed because fetching a tarball from GitHub did not work as expected). Please try resubmitting the patch and let's see if it works this time.

**Comment by chuan9** [comment]
@rakuco got it, I am a new comer, thank you for your help!

**Comment by rakuco** [comment]
> I am a new comer

Welcome aboard! :-)

**Comment by huningxin** [comment]
Thanks for the PR. It passes the bots. I will review ASAP.

File: src/objects.h [hunk 1]
```diff
@@ -6841,6 +6841,10 @@ class Script: public Struct {
   V(SIMD.float32x4, fromInt32x4, Int32x4ToFloat32x4, Float32x4, Int32x4)       \
   V(SIMD.float32x4, fromInt32x4Bits, Int32x4BitsToFloat32x4, Float32x4,        \
     Int32x4)                                                                   \
+  V(SIMD.float32x4, fromFloat64x2, Float64x2ToFloat32x4, Float32x4,            \
+    Float64x2)                                                                 \
+  V(SIMD.float32x4, fromFloat64x2Bits, Float64x2BitsToFloat32x4, Float32x4,    \
+    Float64x2)                                                                 \
```
Comment (huningxin) [inline src/objects.h]: No need for the runtime implementation. Please put this into your CrankShaft PR.

File: src/runtime/runtime-simd.cc [hunk 2]
```diff
@@ -301,8 +301,13 @@ SIMD128_UNARY_FUNCTIONS(DECLARE_SIMD_UNARY_FUNCTION)
 
 
 template<typename T1, typename T2>
-inline void BitsTo(T1 s, T2* t) {
-  memcpy(t, &s, sizeof(T2));
+inline void BitsTo(T1 *s, T2* t) {
+  memcpy(t, s, 1);
```
Comment (huningxin) [inline src/runtime/runtime-simd.cc]: Why change this? Only copy 1 bytes?

File: src/runtime/runtime-simd.cc [hunk 3]
```diff
@@ -301,8 +301,13 @@ SIMD128_UNARY_FUNCTIONS(DECLARE_SIMD_UNARY_FUNCTION)
...
+}
```
Comment (huningxin) [inline src/runtime/runtime-simd.cc]: What's the purpose of this function?

File: src/runtime/runtime-simd.cc [hunk 4]
```diff
@@ -340,10 +376,29 @@ RUNTIME_FUNCTION(                               \
   CONVERT_ARG_CHECKED(SOURCE_TYPE, a, 0);                     \
                                                               \
   TARGET_TYPE::value_t result;                                \
-  for (int i = 0; i < SOURCE_TYPE::kLanes; i++) {             \
-    FUNCTION(a->getAt(i), &result.storage[i]);                \
+  if (#FUNCTION  == "BitsTo") {                               \
+    char *t1 = reinterpret_cast<char*>(&(a->get().storage[0]));\
```
Comment (huningxin) [inline src/runtime/runtime-simd.cc]: Please keep the line end aligned.

File: src/runtime/runtime-simd.cc [hunk 5]
```diff
@@ -340,10 +376,29 @@ RUNTIME_FUNCTION(                               \
...
+  } else {                                                    \
+    if (SOURCE_TYPE::kLanes > TARGET_TYPE::kLanes) {          \
+      for (int i = 0; i < TARGET_TYPE::kLanes; i++) {         \
+        FUNCTION(a->getAt(i), &result.storage[i]);            \
+      }                                                       \
+    } else {                                                  \
+      for (int i = 0; i < SOURCE_TYPE::kLanes; i++) {         \
+        FUNCTION(a->getAt(i), &result.storage[i]);            \
+      }                                                       \
+        int j = TARGET_TYPE::kLanes - SOURCE_TYPE::kLanes;    \
```
Comment (huningxin) [inline src/runtime/runtime-simd.cc]: Keep indent.

File: src/runtime/runtime-simd.cc [hunk 6]
```diff
@@ -340,10 +376,29 @@ RUNTIME_FUNCTION(                               \
...
+          if (j > 0) {                                        \
```
Comment (huningxin) [inline src/runtime/runtime-simd.cc]: ditto.

File: test/mjsunit/simd/float32x4.js [hunk 7]
```diff
@@ -973,3 +972,87 @@ testFloat32ArrayByteOffset();
...
+
+  assertEquals(9.0, n.x);
+  assertEquals(10.0, n.y);
+  assertEquals(11.0, n.z);
+  assertEquals(12.0, n.w);
+};
+
+testFloat32x4fromInt32x4();
+testFloat32x4fromInt32x4();
+%OptimizeFunctionOnNextCall(testFloat32x4fromInt32x4);
```
Comment (huningxin) [inline test/mjsunit/simd/float32x4.js]: As there is no CrankShaft implementation, leave the OptimizeFunctionOnNextCall in your next patch.

File: src/runtime/runtime-simd.cc [hunk 8]
```diff
@@ -340,10 +376,29 @@ RUNTIME_FUNCTION(                               \
   CONVERT_ARG_CHECKED(SOURCE_TYPE, a, 0);                     \
                                                               \
   TARGET_TYPE::value_t result;                                \
-  for (int i = 0; i < SOURCE_TYPE::kLanes; i++) {             \
-    FUNCTION(a->getAt(i), &result.storage[i]);                \
+  if (#FUNCTION  == "BitsTo") {                               \
+    char *t1 = reinterpret_cast<char*>(&(a->get().storage[0]));\
+    char *t2 = reinterpret_cast<char*>(&result.storage[0]);   \
+    for (int i = 0; i < 16; i++) {                            \
```
Comment (huningxin) [inline src/runtime/runtime-simd.cc]: Define the 16 as static const.

File: src/runtime/runtime-simd.cc [hunk 9]
```diff
@@ -340,10 +376,29 @@ RUNTIME_FUNCTION(                               \
...
+      FUNCTION(t1++, t2++);                                   \
+    }                                                         \
+  } else {                                                    \
+    if (SOURCE_TYPE::kLanes > TARGET_TYPE::kLanes) {          \
+      for (int i = 0; i < TARGET_TYPE::kLanes; i++) {         \
+        FUNCTION(a->getAt(i), &result.storage[i]);            \
+      }                                                       \
+    } else {                                                  \
+      for (int i = 0; i < SOURCE_TYPE::kLanes; i++) {         \
+        FUNCTION(a->getAt(i), &result.storage[i]);            \
```
Comment (huningxin) [inline src/runtime/runtime-simd.cc]: Avoid this duplicated function by getting the number of iterations before hand.

File: src/runtime/runtime-simd.cc (hunk 3 above)
Comment (chuan9) [inline src/runtime/runtime-simd.cc]: implement bits copy, when the function SIMD.**.from.**Bits is invoked, this function will worked, the order function is abandoned. I will revise this function to reduce the iterations times.

File: test/mjsunit/simd/float32x4.js (hunk 7 above)
Comment (chuan9) [inline test/mjsunit/simd/float32x4.js]: @huningxin I found that the function was implemented already in crankshaft.

File: src/runtime/runtime-simd.cc (hunk 9 above)
Comment (chuan9) [inline src/runtime/runtime-simd.cc]: yes, I will reduce the iterations by getting the bytes of source type. And add a empty function to pass compilation.

Comment (chuan9) [inline src/runtime/runtime-simd.cc]: yes, I will reduce the iterations by getting the bytes of source type. And add a empty function to pass compilation.

File: test/mjsunit/simd/float32x4.js (hunk 7 above)
Comment (huningxin) [inline test/mjsunit/simd/float32x4.js]: That's fine. Thanks for letting me know.

**Comment by huningxin** [comment]
@chuan9 , you need to rebase your PR.

**Pull request closed without merging**